[web] Use Unified Resources as the default view

avatus opened the pull request:
Closes https://github.com/gravitational/teleport/issues/30597
Updates https://github.com/gravitational/teleport/issues/27462

e: https://github.com/gravitational/teleport.e/pull/2015
This will automatically push users to a legacy resources view if their leaf cluster doesn't support unified resources

If a user wants to disable the view, they can use
```localStorage.setItem("grv_teleport_unified_resources_disabled", "true")```

This varies slightly from the flag we had before (enabled to disabled)

Changelog: Add a new Resources screen instead of a separate list per resource type. Fall back to legacy view on leaf clusters that don't support it just yet.

bl-nero commented on web/packages/teleport/src/TopBar/TopBar.tsx:
```diff
@@ -109,6 +109,38 @@ export function TopBar() {
+    // 2. set this variable conditionally render the "legacy" navigation
+    // When we switch clusters (to leaf or root), we remove the item and perform the check again by pushing
+    // to the resource (new default view).
+    window.localStorage.removeItem(KeysEnum.UNIFIED_RESOURCES_DISABLED);
```
If we simply remove this item, we get rid of an escape hatch to allow customers to get back to the old view if they encounter a problem with the new one and they want to use the legacy until we implement the fix. Perhaps we could do it differently: instead of simply manipulating the global flag, store a set of clusters for which the unified resources feature is disabled, and combine it with this global flag? I.e. if there's no global flag (or it's set to `'false'`), we show the new UI. If we are on a cluster that we detected not to support the new view (using a key-value relationship in local storage), we don't show the new UI, but only if the global flag is not set. And here, instead of removing the global flag, we simply remove given cluster from the list of clusters for which we keep these disabled.

I think this way we would much the release much safer for us and decrease potential for negative customer experience.

avatus replied:
I think we can split it into two and get the same effect. `UNIFIED_RESOURCES_NOT_SUPPORTED` and `UNIFIED_RESOURCES_DISABLED`. that way they are both explicit in what they do

bl-nero commented:
This PR changes the default behavior of our system, and should count as the first PR that actually introduces visible UI changes. Therefore, it should have a changelog label. I suggest the following description:

Changelog: Add a new Resources screen instead of a separate list per resource type. Fall back to legacy view on leaf clusters that don't support it just yet.

kimlisa commented:
why did we remove device trust recommendation?

and also, perhaps i did something wrong, but when i built with this PR and ran it, this is what i got 🤔 

<img width="1710" alt="image" src="https://github.com/gravitational/teleport/assets/43280172/e0b97a36-62d5-4dcf-b6ff-40c78a7993a1">

avatus commented:
Will fix. I realize I have a bad branch. Thank you

avatus committed:
Add legacy leaf cluster support

avatus commented:
fixed the erroneous removal of device trust stuff, also added an `e` counterpart:
https://github.com/gravitational/teleport.e/pull/2015

avatus commented:
> @avatus - this PR will require admin approval to merge due to its size. Consider breaking it up into a series smaller changes

this was an accidental push. local/remote shenanigans and all that

kimlisa reviewed (commented):
I think we need to handle the case where user `hits the back button` from legacy view -> unified view and vice versa

there is also a minified error. i think this was when i was going from leaf view, and hitting back to unified view
<img width="1274" alt="image" src="https://github.com/gravitational/teleport/assets/43280172/6d1d8c8f-0f7c-4189-93e9-3f5d3c9c4ed3">

you also get the minified error when you do this in the dev console and refresh browser: `localStorage.setItem("grv_teleport_unified_resources_disabled", "true")`

avatus committed:
Push to unified if unified resources are enabled

avatus commented:
Caught the error. I added an empty `return` after a history.push in the `Resources` when it wasn't needed. so it was receiving an empty component which is a nono. thank you!

Also, what do you think about my solution to the "back" issue. SImilar to how we do "if resources are not enabled, push to servers", i just added "if they ARE enabled, push to resources". so you can go to a cluster and back and it works fine now. Thoughts @kimlisa ?

avatus committed:
Fix storybook tests
